### PR TITLE
[WIP] Support fraction type

### PR DESCRIPTION
**Pull request by wmedrano**
# Current Status

Some numerical operations were updated to support `FractV` where the numerator and denominator must fit within an `i32`.

Demo:
```lisp
λ > (define x (/ 4))
λ > (+ x 2)
=> 9/4
λ > (* x 2)
=> 1/2
λ > (* x 4)
=> 1
λ > (round x)
=> 0
λ > (exact->inexact x)
=> 0.25
λ > 
```

# Next Steps

- Add support in lexer/parser? Also find out where that is.
- Support `BigFractV`.
- Find all operations that operate on numbers and possibly update with fraction logic.
- Determine if benchmarking is appropriate and do so if required.

### CODE REVIEW

File: crates/steel-core/src/primitives/nums.rs [hunk 1]
```diff
@@ -1,268 +1,213 @@
...
+    };
+    match &args {
+        [] => stop!(ArityMismatch => "/ requires at least one argument"),
+        [x] => recip(x),
+        // TODO: Provide custom implementation to optimize by joining the multiply and recip calls.
+        [x, y] => multiply_2_impl(x, &recip(y)?),
+        [x, ys @ ..] => {
+            let d = multiply_primitive_impl(ys)?;
+            multiply_2_impl(&x, &d)
+        }
```
Comment (wmedrano) [inline crates/steel-core/src/primitives/nums.rs:117]: This should be multiplied by reciprocal of d

**Comment by wmedrano** [comment]
# Current Status

- `FractV` and `BigFract` variants added to `SteelVal`.
- Arithmetic functions that work on `BigNum` now also work on fractions.
- Some unit tests added.

# Next Steps

- Determine if benchmarking is appropriate. I'd prefer to actually check this in now and optimize in subsequent PRs.
- Add more unit tests.
- Add parser/lexer support.

# Overall Sentiment

- Nice to have number support in line with other Schemes.
- I don't like how complicated some of the arithmetic code got, but I guess its necessary.

File: crates/steel-core/src/steel_vm/primitives.rs [hunk 2]
```diff
@@ -894,11 +951,17 @@ fn expt(left: &SteelVal, right: &SteelVal) -> Result<SteelVal> {
 #[steel_derive::function(name = "exp", constant = true)]
 fn exp(left: &SteelVal) -> Result<SteelVal> {
     match left {
-        SteelVal::IntV(i) if *i == 0 => Ok(SteelVal::IntV(1)),
+        SteelVal::IntV(0) => Ok(SteelVal::IntV(1)),
```
Comment (wmedrano) [inline crates/steel-core/src/steel_vm/primitives.rs:954]: Kind of interesting. Guile gives `1` for this, but Racket gives `1.0`.

Comment (mattwparas) [reply crates/steel-core/src/steel_vm/primitives.rs:953]: Well I was definitely trying to replicate racket behavior - I guess also maybe racket tries to convert to exact where possible, but that also might be a completely baseless statement

**Comment by mattwparas** [comment]
> # Current Status
> * `FractV` and `BigFract` variants added to `SteelVal`.
> * Arithmetic functions that work on `BigNum` now also work on fractions.
> * Some unit tests added.
> 
> # Next Steps
> * Determine if benchmarking is appropriate. I'd prefer to actually check this in now and optimize in subsequent PRs.
> * Add more unit tests.
> * Add parser/lexer support.
> 
> # Overall Sentiment
> * Nice to have number support in line with other Schemes.
> * I don't like how complicated some of the arithmetic code got, but I guess its necessary.

This sounds good - so based on this you want to get this merged first before adding more? If so, sounds good - I can point you to the parser and lexer spots if you'd like to do that as well.

**Review by mattwparas — APPROVED** [review]

**Pull request merged**